File: data_fetching/pumpFunDataFetching.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

const apiKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJub25jZSI6IjVjZDZjZDk5LTg5MjUtNGRiYS1iZWQyLWJjNzNkYzQ3ZTVmYSIsIm9yZ0lkIjoiNDU5ODY2IiwidXNlcklkIjoiNDczMTE4IiwidHlwZUlkIjoiOTVjZGQwYmMtNDViYi00MWVhLWEwMGItMWRhODZjZDNiM2Q2IiwidHlwZSI6IlBST0pFQ1QiLCJpYXQiOjE3NTI3NjcyNjUsImV4cCI6NDkwODUyNzI2NX0.BOMHcWhyTzWhnT3an-pporsYhIGm0yGyCO2CZ381liY"
const baseURL = "https://solana-gateway.moralis.io"

// Set this to the token mint address you want to test
var tokenMint = "CzLSujWBLFsSjncfkh59rUFqvafWcY5tzedWJSuypump" // Using a token from the actual response
const isFraudLabel = 0                                         // 1 = Fraud, 0 = Legit


// Structs

// Metadata
type MetadataResponse struct {
	Mint                 string `json:"mint"`
	Name                 string `json:"name"`
	Symbol               string `json:"symbol"`
	Decimals             string `json:"decimals"`
	TotalSupply          string `json:"totalSupply"`
	TotalSupplyFormatted string `json:"totalSupplyFormatted"`
	Metaplex             struct {
		SellerFeeBasisPoints int    `json:"sellerFeeBasisPoints"`
		IsMutable            bool   `json:"isMutable"`
		PrimarySaleHappened  int    `json:"primarySaleHappened"`
		UpdateAuthority      string `json:"updateAuthority"`
	} `json:"metaplex"`
}

// Price
type PriceResponse struct {
	UsdPrice    float64 `json:"usdPrice"`
	NativePrice struct {
		Value string `json:"value"`
	} `json:"nativePrice"`
	ExchangeName    string `json:"exchangeName"`
	ExchangeAddress string `json:"exchangeAddress"`
}

// Swap
type SwapsResponse struct {
	Result []struct {
		TransactionType string    `json:"transactionType"`
		BlockTimestamp  time.Time `json:"blockTimestamp"`
		WalletAddress   string    `json:"walletAddress"`
		TotalValueUsd   float64   `json:"totalValueUsd"`
	} `json:"result"`
}

// Unified Schema
type UnifiedTokenData struct {
	Mint                 string `json:"mint"`
	Name                 string `json:"name"`
	Symbol               string `json:"symbol"`
	Decimals             string `json:"decimals"`
	TotalSupply          string `json:"totalSupply"`
	TotalSupplyFormatted string `json:"totalSupplyFormatted"`
	SellerFeeBasisPoints int    `json:"sellerFeeBasisPoints"`
	IsMutable            bool   `json:"isMutable"`
	PrimarySaleHappened  bool   `json:"primarySaleHappened"`
	UpdateAuthority      string `json:"updateAuthority"`

	UsdPrice         float64 `json:"usdPrice"`
	NativePriceValue string  `json:"nativePriceValue"`
	ExchangeName     string  `json:"exchangeName"`
	ExchangeAddress  string  `json:"exchangeAddress"`

	FirstSwapTimestamp time.Time `json:"firstSwapTimestamp"`
	FirstSwapType      string    `json:"firstSwapType"`
	FirstBuyerOrSeller string    `json:"firstBuyerOrSeller"`
	FirstTradeValueUsd float64   `json:"firstTradeValueUsd"`

	Label int `json:"label"` // Added label field
}

func fetchJSON(url string, result interface{}) error {
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Add("X-API-Key", apiKey)
	req.Header.Add("accept", "application/json")
	client := &http.Client{}
	res, err := client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	body, _ := ioutil.ReadAll(res.Body)
	return json.Unmarshal(body, result)
}

func main() {
	token := tokenMint

	// Fetch metadata
	meta := MetadataResponse{}
	err := fetchJSON(fmt.Sprintf("%s/token/mainnet/%s/metadata", baseURL, token), &meta)
	if err != nil {
		log.Fatal("Metadata fetch failed:", err)
	}

	// Fetch price
	price := PriceResponse{}
	err = fetchJSON(fmt.Sprintf("%s/token/mainnet/%s/price", baseURL, token), &price)
	if err != nil {
		log.Fatal("Price fetch failed:", err)
	}

	// Fetch swaps
	swaps := SwapsResponse{}
	err = fetchJSON(fmt.Sprintf("%s/token/mainnet/%s/swaps?order=ASC&limit=1", baseURL, token), &swaps)
	if err != nil {
		log.Fatal("Swap fetch failed:", err)
	}
	var firstSwapTime time.Time
	var firstBuyerOrSeller string
	var firstType string
	var firstTotalUSD float64
	if len(swaps.Result) > 0 {
		first := swaps.Result[0] // Oldest
		firstSwapTime = first.BlockTimestamp
		firstBuyerOrSeller = first.WalletAddress
		firstType = first.TransactionType
		firstTotalUSD = first.TotalValueUsd
	}

	// Aggregate
	aggregated := UnifiedTokenData{
		Mint:                 meta.Mint,
		Name:                 meta.Name,
		Symbol:               meta.Symbol,
		Decimals:             meta.Decimals,
		TotalSupply:          meta.TotalSupply,
		TotalSupplyFormatted: meta.TotalSupplyFormatted,
		SellerFeeBasisPoints: meta.Metaplex.SellerFeeBasisPoints,
		IsMutable:            meta.Metaplex.IsMutable,
		PrimarySaleHappened:  meta.Metaplex.PrimarySaleHappened == 1,
		UpdateAuthority:      meta.Metaplex.UpdateAuthority,
		UsdPrice:             price.UsdPrice,
		NativePriceValue:     price.NativePrice.Value,
		ExchangeName:         price.ExchangeName,
		ExchangeAddress:      price.ExchangeAddress,
		FirstSwapTimestamp:   firstSwapTime,
		FirstSwapType:        firstType,
		FirstBuyerOrSeller:   firstBuyerOrSeller,
		FirstTradeValueUsd:   firstTotalUSD,
		Label:                isFraudLabel, // Set label based on fraud status
	}

	jsonData, err := json.MarshalIndent(aggregated, "", "  ")
	if err != nil {
		log.Fatal("JSON marshaling failed:", err)
	}

	// Create json_output directory if it doesn't exist
	outputDir := "json_output"
	err = os.MkdirAll(outputDir, 0755)
	if err != nil {
		log.Fatal("Directory creation failed:", err)
	}

	// Create filename with timestamp
	filename := fmt.Sprintf("token_data_%s_%d.json", token, time.Now().Unix())
	filePath := filepath.Join(outputDir, filename)

	// Write to file
	err = ioutil.WriteFile(filePath, jsonData, 0644)
	if err != nil {
		log.Fatal("File writing failed:", err)
	}

	fmt.Printf("Token data successfully exported to: %s\n", filePath)
}

File: data_fetching/pumpFunDataFetching.py
import json
import os
import time
import requests
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Any, Optional

API_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJub25jZSI6IjVjZDZjZDk5LTg5MjUtNGRiYS1iZWQyLWJjNzNkYzQ3ZTVmYSIsIm9yZ0lkIjoiNDU5ODY2IiwidXNlcklkIjoiNDczMTE4IiwidHlwZUlkIjoiOTVjZGQwYmMtNDViYi00MWVhLWEwMGItMWRhODZjZDNiM2Q2IiwidHlwZSI6IlBST0pFQ1QiLCJpYXQiOjE3NTI3NjcyNjUsImV4cCI6NDkwODUyNzI2NX0.BOMHcWhyTzWhnT3an-pporsYhIGm0yGyCO2CZ381liY"
BASE_URL = "https://solana-gateway.moralis.io"

# Set this to the token mint address you want to test
TOKEN_MINT = "CzLSujWBLFsSjncfkh59rUFqvafWcY5tzedWJSuypump"  # Using a token from the actual response
IS_FRAUD_LABEL = 0  # 1 = Fraud, 0 = Legit


class MetadataResponse:
    def __init__(self, data: Dict[str, Any]):
        self.mint = data.get("mint", "")
        self.name = data.get("name", "")
        self.symbol = data.get("symbol", "")
        self.decimals = data.get("decimals", "")
        self.total_supply = data.get("totalSupply", "")
        self.total_supply_formatted = data.get("totalSupplyFormatted", "")
        
        metaplex = data.get("metaplex", {})
        self.seller_fee_basis_points = metaplex.get("sellerFeeBasisPoints", 0)
        self.is_mutable = metaplex.get("isMutable", False)
        self.primary_sale_happened = metaplex.get("primarySaleHappened", 0)
        self.update_authority = metaplex.get("updateAuthority", "")


class PriceResponse:
    def __init__(self, data: Dict[str, Any]):
        self.usd_price = data.get("usdPrice", 0.0)
        native_price = data.get("nativePrice", {})
        self.native_price_value = native_price.get("value", "")
        self.exchange_name = data.get("exchangeName", "")
        self.exchange_address = data.get("exchangeAddress", "")


class SwapResult:
    def __init__(self, data: Dict[str, Any]):
        self.transaction_type = data.get("transactionType", "")
        # Parse ISO timestamp to datetime
        timestamp_str = data.get("blockTimestamp", "")
        if timestamp_str:
            self.block_timestamp = datetime.fromisoformat(timestamp_str.replace('Z', '+00:00'))
        else:
            self.block_timestamp = datetime.now()
        self.wallet_address = data.get("walletAddress", "")
        self.total_value_usd = data.get("totalValueUsd", 0.0)


class SwapsResponse:
    def __init__(self, data: Dict[str, Any]):
        result_data = data.get("result", [])
        self.result = [SwapResult(item) for item in result_data]


class UnifiedTokenData:
    def __init__(self):
        self.mint = ""
        self.name = ""
        self.symbol = ""
        self.decimals = ""
        self.total_supply = ""
        self.total_supply_formatted = ""
        self.seller_fee_basis_points = 0
        self.is_mutable = False
        self.primary_sale_happened = False
        self.update_authority = ""
        
        self.usd_price = 0.0
        self.native_price_value = ""
        self.exchange_name = ""
        self.exchange_address = ""
        
        self.first_swap_timestamp = None
        self.first_swap_type = ""
        self.first_buyer_or_seller = ""
        self.first_trade_value_usd = 0.0
        
        self.label = 0
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "mint": self.mint,
            "name": self.name,
            "symbol": self.symbol,
            "decimals": self.decimals,
            "totalSupply": self.total_supply,
            "totalSupplyFormatted": self.total_supply_formatted,
            "sellerFeeBasisPoints": self.seller_fee_basis_points,
            "isMutable": self.is_mutable,
            "primarySaleHappened": self.primary_sale_happened,
            "updateAuthority": self.update_authority,
            "usdPrice": self.usd_price,
            "nativePriceValue": self.native_price_value,
            "exchangeName": self.exchange_name,
            "exchangeAddress": self.exchange_address,
            "firstSwapTimestamp": self.first_swap_timestamp.isoformat() if self.first_swap_timestamp else "",
            "firstSwapType": self.first_swap_type,
            "firstBuyerOrSeller": self.first_buyer_or_seller,
            "firstTradeValueUsd": self.first_trade_value_usd,
            "label": self.label
        }


def check_for_duplicate_mint(mint: str) -> None:
    """Check if the given mint address already exists in any JSON file"""
    output_dir = Path("json_output")
    
    # Check if directory exists
    if not output_dir.exists():
        # Directory doesn't exist, no duplicates possible
        return
    
    # Check each JSON file
    for file_path in output_dir.glob("*.json"):
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                token_data = json.load(f)
            
            # Check if mint matches
            if token_data.get("mint") == mint:
                raise Exception(f"duplicate mint address detected: {mint} already exists in file: {file_path}")
                
        except json.JSONDecodeError:
            print(f"Warning: failed to parse JSON in file {file_path}")
            continue
        except FileNotFoundError:
            print(f"Warning: failed to read file {file_path}")
            continue


def fetch_json(url: str) -> Dict[str, Any]:
    """Fetch JSON data from the given URL with API key authentication"""
    headers = {
        "X-API-Key": API_KEY,
        "accept": "application/json"
    }
    
    response = requests.get(url, headers=headers)
    response.raise_for_status()
    return response.json()


def main():
    token = TOKEN_MINT
    
    # Check for duplicate mint address before processing
    try:
        check_for_duplicate_mint(token)
    except Exception as e:
        print(f"Duplicate check failed: {e}")
        return
    
    try:
        # Fetch metadata
        metadata_url = f"{BASE_URL}/token/mainnet/{token}/metadata"
        metadata_data = fetch_json(metadata_url)
        meta = MetadataResponse(metadata_data)
        
        # Fetch price
        price_url = f"{BASE_URL}/token/mainnet/{token}/price"
        price_data = fetch_json(price_url)
        price = PriceResponse(price_data)
        
        # Fetch swaps
        swaps_url = f"{BASE_URL}/token/mainnet/{token}/swaps?order=ASC&limit=1"
        swaps_data = fetch_json(swaps_url)
        swaps = SwapsResponse(swaps_data)
        
        # Initialize swap variables
        first_swap_time = None
        first_buyer_or_seller = ""
        first_type = ""
        first_total_usd = 0.0
        
        if len(swaps.result) > 0:
            first = swaps.result[0]  # Oldest
            first_swap_time = first.block_timestamp
            first_buyer_or_seller = first.wallet_address
            first_type = first.transaction_type
            first_total_usd = first.total_value_usd
        
        # Aggregate
        aggregated = UnifiedTokenData()
        aggregated.mint = meta.mint
        aggregated.name = meta.name
        aggregated.symbol = meta.symbol
        aggregated.decimals = meta.decimals
        aggregated.total_supply = meta.total_supply
        aggregated.total_supply_formatted = meta.total_supply_formatted
        aggregated.seller_fee_basis_points = meta.seller_fee_basis_points
        aggregated.is_mutable = meta.is_mutable
        aggregated.primary_sale_happened = meta.primary_sale_happened == 1
        aggregated.update_authority = meta.update_authority
        aggregated.usd_price = price.usd_price
        aggregated.native_price_value = price.native_price_value
        aggregated.exchange_name = price.exchange_name
        aggregated.exchange_address = price.exchange_address
        aggregated.first_swap_timestamp = first_swap_time
        aggregated.first_swap_type = first_type
        aggregated.first_buyer_or_seller = first_buyer_or_seller
        aggregated.first_trade_value_usd = first_total_usd
        aggregated.label = IS_FRAUD_LABEL  # Set label based on fraud status
        
        # Convert to JSON
        json_data = json.dumps(aggregated.to_dict(), indent=2)
        
        # Create json_output directory if it doesn't exist
        output_dir = Path("json_output")
        output_dir.mkdir(exist_ok=True)
        
        # Create filename with timestamp
        timestamp = int(time.time())
        filename = f"token_data_{token}_{timestamp}.json"
        file_path = output_dir / filename
        
        # Write to file
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(json_data)
        
        print(f"Token data successfully exported to: {file_path}")
        
    except requests.RequestException as e:
        print(f"API request failed: {e}")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()
